game/procurement.py

```diff
@@ -116,52 +128,236 @@ def repair_runways(self, budget: float) -> float:
+
+        if faction_has_ifv_access:
+            if costs_of_all_vehicles is 0:
+                prioritylist.append(
+                    (
+                        TYPE_IFV,
+                        self.game.settings.ground_forces_procurement_ratio.ifv_ratio,
+                    )
+                )
+            else:
+                current_ratio = ifv_costs / costs_of_all_vehicles
+                if (
+                    int(current_ratio)
+                    < self.game.settings.ground_forces_procurement_ratio.ifv_ratio
+                ):
+                    prioritylist.append((TYPE_IFV, current_ratio))
+
+        if faction_has_apc_access:
+            if costs_of_all_vehicles is 0:
+                prioritylist.append(
+                    (
+                        TYPE_APC,
+                        self.game.settings.ground_forces_procurement_ratio.apc_ratio,
+                    )
+                )
+            else:
+                current_ratio = apc_costs / costs_of_all_vehicles
+                if (
+                    int(current_ratio)
+                    < self.game.settings.ground_forces_procurement_ratio.apc_ratio
+                ):
+                    prioritylist.append((TYPE_APC, current_ratio))
+
+        if faction_has_artillery_access:
+            if costs_of_all_vehicles is 0:
+                prioritylist.append(
+                    (
+                        TYPE_ARTILLERY,
+                        self.game.settings.ground_forces_procurement_ratio.artillery_ratio,
+                    )
+                )
+            else:
+                current_ratio = artillery_costs / costs_of_all_vehicles
+                if (
+                    int(current_ratio)
+                    < self.game.settings.ground_forces_procurement_ratio.artillery_ratio
+                ):
+                    prioritylist.append((TYPE_ARTILLERY, current_ratio))
+
+        if faction_has_shorad_access:
+            if costs_of_all_vehicles is 0:
+                prioritylist.append(
+                    (
+                        TYPE_SHORAD,
+                        self.game.settings.ground_forces_procurement_ratio.shorad_ratio,
+                    )
+                )
+            else:
+                current_ratio = shorad_costs / costs_of_all_vehicles
+                if (
+                    int(current_ratio)
+                    < self.game.settings.ground_forces_procurement_ratio.shorad_ratio
+                ):
+                    prioritylist.append((TYPE_SHORAD, current_ratio))
+
+        return prioritylist
+
+    def calculate_investment_of_vehicleType(
+        self, cp: ControlPoint, vehicle_type: list
+    ) -> int:
+        costs = 0
+        vehicles_in_base = self.filter.get_all_vehicletype_from_dic(
+            cp.base.armor, vehicle_type
+        )
+        vehicles_ordered = self.filter.get_all_vehicletype_from_dic(
+            cp.pending_unit_deliveries.units, vehicle_type
+        )
```

Comment:
This is incorrect (as was my snippet in the previous review cycle, fwiw). It does not count any of the units that have spawned and are in transit. You need a variant of `total_ground_units_allocated_to` (and by extension a variant of `ControlPoint.expected_ground_units_next_turn`).

The "static" class (python doesn't have those in the same form as Java, but the pattern you were looking for was `@classmethod`)  you're using here seems like the wrong home for these things. Seems like everything that class does should be `ControlPoint.allocated_ground_units_of_class(game: Game, unit_class: GroundUnitClass) -> int`

wdyt?